### ISSUE TITLE
Option to allow for the recipient to delete the document.

### ISSUE THREAD

**Issue by dominique-p** [PrivateBin/PrivateBin#782]
Request for Enhancement:
Have an option to allow for the recipient to delete the document.
A bit like with "Burn after reading", but allows for multiple downloads and deletion under the control of the recipient (before the expiration time).

Regards,
Dominique

**Comment by rugk** [comment]
What is the exact use case for that?
What do you try to achieve?

—

Also technically, I guess this is quite hard to implement, as we already generated and saved the paste when we generate the delete link AFAIK.
Though, of course one may just generate the random stuff client-side and then send it to the server… and thus it could maybe be included in the paste/encrypted data.

**Comment by dominique-p** [comment]
Sometimes, when "Burn after reading" is set for a document, the recipient makes a mistakes and loses access to the document, which has to be sent again. But there is  also the suspicion that somebody else has accessed the document. Some antispam tools do lookup the content of URLs, which would prevent the recipient to access it. Without "Burn after reading" once the document is received, it remains available until the expiration which could be a relatively long time. Enabling the document deletion by the recipient would ensure that he has ascertained that he received and processed (saved, memorized, etc.) it and that it would not stay available any longer.  RFE 2: A counter of how many times a document has been accessed would also enable the recipient to know if a third party would have accessed it, raising an alert that some "man in the middle" is spying. 
It's the difference between autoburn "Mission: impossible" style and actually putting the document in the fire yourself once you memorized its content.
Now, it is not critical and might not be worth the extra complexity. 
Thanks for your attention anyway!

Regards,
Dominique

**Comment by rugk** [comment]
> Some antispam tools do lookup the content of URLs, which would prevent the recipient to access it. 

We have measures implemented to stop such bots. If you find one, which breaks a burn-after-read paste, please open a new issue with details on what software does so (preferably including the user agent).

**Comment by elrido** [comment]
>  Have an option to allow for the recipient to delete the document.

PrivateBin has no concept of a recipient, only of viewers and creators. The deletion link is presented to the creator along side the link to view the paste. It is up to the creator to share either or both links with a recipient and how they do that.

There are some measures in place that should prevent the deletion of "burn after reading" paste by such a tool. It would need to load and execute all of the javascript libraries and WASM modules and implement the WebCrypto API like a full web browser and perform the API request that actually retrieves the paste in question. Just calling the URL of the view link on it's own wouldn't be sufficient to fetch/delete the paste.

On the other hand, the deletion link isn't protected in this way and if such an anti-spam tool looks up the deletion links, it would trigger the deletion instantly.

**Comment by modulec** [comment]
> > The deletion link is presented to the creator along side the link to view the paste. It is up to the creator to share either or both links with a recipient and how they do that.
> 
> Of course, that is likely the easiest way to solve the use case.
> If you want to avoid sending two links or the mentioned problem if antispam tools trigger the deletion link, you can also enable the comment feature in a paste and then just paste the deletion link in there…

Tnx for this workaround, works fine for now.

I also do agree with the @dominique-p in his explanation of the usecase in https://github.com/PrivateBin/PrivateBin/issues/782#issuecomment-824456262 

We found that sometimes bots or IM's will often trigger first page hit often burning the secret, and it would be great to have a pre-build button on secret itself that will delete it on button press.

I will try to find the list of user agents through the server logs and update you with the list.

Even better maybe would be having a button displayed on the secret page before displaying the secret to the user, informing the user that after he presses the button, secret would be displayed and then burned.

This should also solve the issue of having to maintain the bot list as the button would be an effective counter for bots if the actual action is done only in valid browsers. 

Viewed counter in this case is also not needed as the secret is only viewable once.

**Comment by rugk** [comment]
> We found that sometimes bots or IM's will often trigger first page hit often burning the secret,

Again, BTW, if you find such an issue, please do definitively open a new issue, because that is a bug. We have measures to stop bots from accidentally opening a paste in place, currently we detect them via their user agent, but we could do anything more elaborate if required.

So if you find such one IM or bot, please open a new issue for each of them and include the user-agent, if possible. That would help a lot, thanks.

**Comment by rugk** [comment]
> Even better maybe would be having a button displayed on the secret page before displaying the secret to the user, informing the user that after he presses the button, secret would be displayed and then burned.

That has also been requested in the past e.g. in https://github.com/PrivateBin/PrivateBin/issues/529 and https://github.com/PrivateBin/PrivateBin/issues/347

**Comment by HLeithner** [comment]
Hmm, a limitation on views would also be useful against abusing the systems by bot networks. I have a bin with 700k requests last month.

**Comment by rugk** [comment]
@HLeithner [As said above](https://github.com/PrivateBin/PrivateBin/issues/782#issuecomment-932629527), bots should be stopped from opening the paste altogether and we have mechanisms to prevent that. Please open an issue with details about the bot that accesses am paste.

I also don't see why a counter would be more useful than and technically it is also a totally separate feature request. Also see https://github.com/PrivateBin/PrivateBin/issues/800 as a very similar request, which e.g. explains that we of course do want to collect as few usage data as possible. Counting paste visits is a thing we thus explicitly want to avoid.

**Comment by HLeithner** [comment]
@rugk in this case it's actually not a bot that access the bin, it seems to be a documentation how to by pass law enforcement or ISP block to access a specific website by using VPN or alternative DNS servers. So the access are by real users and not by a bot.

**Comment by rugk** [comment]
Okay, but then it's a legitimate access and no “abusing the systems by bot networks”, though I don't understand how that thing works. In any case, then I see no reason to limit access in any way. :upside_down_face:

**Comment by HLeithner** [comment]
at the time of writing I thought it was a bot... depends if you like to host bins which unlimited requests if you provide a service which is defined by concept to be accessed by a limited amount of people.

At the moment the bin in question is listed at google incl. key. If I had a problem with this bin I would had deleted it already. But limiting a "private" bin to for example to 100k requests would fit the needs of 99% more or less. At least for me it would be enough as long as it's clearly communicated to the creator of the bin (like the time restriction)

**Comment by rugk** [comment]
Well that always depends on your definition of "private". Private in our technical way of viewing mostly includes encryption and some other metadata-considerations etc., which are irrelevant here, however. It does not necessarily mean the (size of the) audience is limited or even is below a fixed limit such as 100k requests. Who knows…

So if one wants to host a very public information on PrivateBin, I see no reason, why we would want to prevent it. And as it seems to have occurred for you, actually, in your case people found a reason to reach a bigger target audience with apparently leads to many requests to one paste. I see nothing fundamentally wrong with that, especially if it has such a good intend as your one of providing and spreading anti-censorship material – which e.g. could be blocked on other websites or if it was too widely available (just a random theory).
Actually, [one of the reasons we communicate in our FAQ for providing a PrivateBin instance](https://github.com/PrivateBin/PrivateBin/wiki/FAQ#why-should-i-use-privatebin) is [to provide a service for whose, who need it](https://www.schneier.com/blog/archives/2015/06/why_we_encrypt.html), made possible by encryption itself.

Also please note that [PrivateBin does have an integrated rate/traffic limiter](https://github.com/PrivateBin/PrivateBin/wiki/Configuration#trafficlimit), so things like spamming an instance should already be prevented.
Note of course if you are not fine with the content of a paste or legally required or whatnot, you can always delete it as a server admin. 

Anyway, we kinda get off-topic, as the actually requested feature in this issue was totally different. If you want to still have a counter feature, feel free to open a new issue, but the arguments presented here, would still need to be discussed there.

**Comment by planetahuevo** [comment]
I would like to suggest what I suggested 2 years ago: 
https://github.com/PrivateBin/PrivateBin/issues/541 which has been marked as duplicated and closed, but I think it has not been suggested here. 

The idea is not to change the way the system works, but only change the frontend for the viewer offering a button tu burn the secret. 

The link already exist and it is shown to the creator. 

My suggestion is to have an option to show that link as button to the viewer as well.

This should be an option to the installation (same as the comments system and other features) and disabled by default. 

This should be much easier than count visits, ask before viewer or delete just after one view. 
It is manual, and the viewer can decide to delete or not once it has been read. 
It will fit some use cases, and not others, therefore a default disabled option seems the best approach.

What do you think?

**Comment by rugk** [comment]
It's not as easy to implement as you think. See the other issue where I've replied in detail.

Also I guess this was what this issue was originally about.

**Comment by rugk** [comment]
Maybe we shoild close this one as iy drifted off-topic and instead leave your one open?

**Comment by planetahuevo** [comment]
> Maybe we should close this one as it drifted off-topic and instead leave your one open?
Yes, I agree you can keep only one, it is all about the same functionality. 
Whatever you pick would be fine with me.

**Comment by rugk** [comment]
Duplicate of #541
